File: py/xrecord.py
#!/usr/bin/python

import sys
import go
import torch
import os.path
from os import listdir
import randmove
import resnet
from mcts import MCTSPlayer

def main(count, path):
    sys.setrecursionlimit(500000)
    go.init(9)
    playerBlack = MCTSPlayer(30, resnet.Policy(40, '../data/r1/resnet_pars.pkl'))
    playerWhite = MCTSPlayer(30, randmove.Policy(40))

    records = [f for f in listdir(path) if f[-4:] == 'json' and f[:6] == 'record']
    fcount = len(records)

    vertex = None
    caps = None

    black_win = 0
    white_win = 0
    draw = 0
    
    c = 1
    while c <= count:
        records = '[\n'
        i = 0

        print('ready: %d in %d, POSPOOL: %d' % (c, count, len(go.POSITION_POOL)))

        while go.POSITION.pass_count() < 2:
            i += 1

            legal = True
            if i % 2 == 1:
                legal = playerBlack.move()
            else:
                legal = playerWhite.move()

            if not legal:
                print('Illegal move!')
                break

            go.POSITION.debug()

            record = go.POSITION.toJSON()
            records += '  '
            records += record

            if go.POSITION.pass_count() < 2:
                records += ',\n'
            else:
                records += '\n]'
        
        score = go.POSITION.score() - go.KOMI
        playerBlack.clear()
        playerWhite.clear()
        go.clear()

        fcount += 1
        filename = path + 'record_%d.json' % fcount

        f = open(filename, 'w')
        f.write(records)
        f.close()

        if score > 0:
            black_win += 1
        elif score < 0:
            white_win += 1
        else:
            draw += 1
        
        c += 1

    print('back win: %d, white win: %d, draw: %d' % (black_win, white_win, draw))

    

if __name__ == '__main__':
    count = 1
    path = '../data/'

    if len(sys.argv) >= 2:
        path += sys.argv[1] + '/'

    if len(sys.argv) >= 3:
        count = int(sys.argv[2])

    main(count, path)

File: cpp/src/xrecord.cpp

#include <string>
#include <iostream>
#include <fstream>
#include <sstream>
#include <chrono>
#include <iomanip>
#include "go.h"
#include "MCTSPlayer.h"
#include "RandMove.h"
#include "rapidjson/document.h"
#include "rapidjson/prettywriter.h"

using namespace std;
using namespace rapidjson;

int main(int argc, char *argv[])
{
    int count = 1;
    string path = "../../data/";

    if (argc >= 2)
    {
        path += argv[1];
        path += "/";
    }

    if (argc >= 3)
    {
        count = atoi(argv[2]);
    }

    go::init();

    MCTSPlayer *playerBlack = new MCTSPlayer(5.f, new RandMove(40));
    MCTSPlayer *playerWhite = new MCTSPlayer(5.f, new RandMove(40));

    int black_win = 0;
    int white_win = 0;

    for (int c = 0; c != count; ++c)
    {
        auto t = chrono::system_clock::to_time_t(chrono::system_clock::now());
        stringstream filename;
        filename << put_time(localtime(&t), "%y%m%d%H%M%S") << ".json";

        cout << "ready: " << c << " in " << count << ", PP:" << go::POSITION_POOL.size()
             << ", GP:" << Group::POOL.size() << ", MP:" << MCTSPlayer::POOL.size()
             << ", File:" << filename.str() << endl;

        Document positions(kArrayType);
        Document::AllocatorType &allocator = positions.GetAllocator();

        MCTSPlayer *player = playerBlack;

        while (go::POSITION->passCount() < 2)
        {
            bool legal = player->move();
            if (!legal)
            {
                cerr << "Illegal move!" << endl;
                break;
            }

            positions.PushBack(go::POSITION->toJSON(allocator), allocator);

            if (player == playerBlack)
            {
                player = playerWhite;
            }
            else
            {
                player = playerBlack;
            }
        }

        float score = go::POSITION->score() - go::KOMI;
        if (score > 0)
        {
            ++black_win;
        }
        else if (score < 0)
        {
            ++white_win;
        }

        playerBlack->clear();
        playerWhite->clear();
        go::clear();

        StringBuffer buffer;
        Writer<StringBuffer> writer(buffer);
        positions.Accept(writer);

        string fullpath = path + filename.str();
        ofstream fs(fullpath);
        if (fs)
        {
            fs << buffer.GetString() << endl;
            fs.close();
        } else {
            cerr << "create file '" << fullpath << "' failed" << endl;
        }
    }

    cout << "black win: " << black_win << ", white win: " << white_win
         << ", draw: " << count - black_win - white_win << endl;
}